### PR TITLE
Deploy 081 - Upgrade Harvester to support Balancer, Curve and Uniswap V3

### PR DESCRIPTION
**Pull request by shahthepro**
Deployment file: `081_upgrade_harvester.js`

Deployed contracts:

| Contract name | Address |
| --------------- | -------- |
| AuraWETHPriceFeed (new) | [0x94e16bC08d7CCd7f2999Eb5eA3f35DD1EDCBd15B](https://etherscan.io/address/0x94e16bC08d7CCd7f2999Eb5eA3f35DD1EDCBd15B) |
| OETHOracleRouter (replaced) | [0x468A68da3cefcDD644ce0Ea9B9564b246218aeeC](https://etherscan.io/address/0x468A68da3cefcDD644ce0Ea9B9564b246218aeeC) |
| Harvester (upgraded) | [0x07a450739cdd8Ac0a59C9d752ce915a86010658B](https://etherscan.io/address/0x07a450739cdd8Ac0a59C9d752ce915a86010658B) |
| OETHHarvester (upgraded) | [0x4De255cd196469752632e76CbAaAEeE0070D9A85](https://etherscan.io/address/0x4De255cd196469752632e76CbAaAEeE0070D9A85) |

Other addresses:

| Name | Address |
| ------ | -------- |
| Deployer | [0x58890a9cb27586e83cb51d2d26bbe18a1a647245](https://etherscan.io/address/0x58890a9cb27586e83cb51d2d26bbe18a1a647245) |
| Proposer | [0x6a6D776120f7e4a8dba5F6bF49b85cb340Cfe241](https://etherscan.io/address/0x6a6D776120f7e4a8dba5F6bF49b85cb340Cfe241) |

Governance Proposal:

| | |
| - | - |
| Proposal ID | [70744121595007528818249644545963691097758184661168820806929451960448344720141](https://governance.ousd.com/proposals/70744121595007528818249644545963691097758184661168820806929451960448344720141) |
| Proposal Tx | [0x84ddb0ca9d15aa8275c9d5eb3d259b9fff974534db274d3937d01f709c420fab](https://etherscan.io/tx/0x84ddb0ca9d15aa8275c9d5eb3d259b9fff974534db274d3937d01f709c420fab) |

Proposal Actions:
- Upgrade HarvesterProxy and OETHHarvesterProxy to new deployed implementation
- Configure OETH Vault to use the newly deployed OETHOracleRouter
- Update reward token configs on both OETH and OUSD Harvesters
- Change OETHDripper drip time to 14 days (from 3 days)

### CODE REVIEW

**Comment by naddison36** [comment]
The verification of the deployed contracts against the commit hash 7b4a2605a6d628795a7c00dc64c1dba2f3604c1d

```
sol2uml diff 0x94e16bC08d7CCd7f2999Eb5eA3f35DD1EDCBd15B .,node_modules --aFile contracts/oracle/AuraWETHPriceFeed.sol
Compared the "contracts/oracle/AuraWETHPriceFeed.sol" file with the "AuraWETHPriceFeed" contract with address 0x94e16bC08d7CCd7f2999Eb5eA3f35DD1EDCBd15B on mainnet
to local files under folders ".,node_modules"

match   @openzeppelin/contracts/utils/math/SafeCast.sol
match   contracts/governance/Governable.sol
match   contracts/governance/Strategizable.sol
match   contracts/interfaces/balancer/IOracleWeightedPool.sol
match   contracts/interfaces/chainlink/AggregatorV3Interface.sol
match   contracts/oracle/AuraWETHPriceFeed.sol

sol2uml diff 0x468A68da3cefcDD644ce0Ea9B9564b246218aeeC .,node_modules --aFile contracts/oracle/OETHOracleRouter.sol
Compared the "contracts/oracle/OETHOracleRouter.sol" file with the "OETHOracleRouter" contract with address 0x468A68da3cefcDD644ce0Ea9B9564b246218aeeC on mainnet
to local files under folders ".,node_modules"

match   @openzeppelin/contracts/utils/math/SafeCast.sol
match   @openzeppelin/contracts/utils/math/SafeMath.sol
match   contracts/interfaces/chainlink/AggregatorV3Interface.sol
match   contracts/interfaces/IBasicToken.sol
match   contracts/interfaces/IOracle.sol
match   contracts/oracle/OETHOracleRouter.sol
match   contracts/oracle/OracleRouterBase.sol
match   contracts/utils/Helpers.sol
match   contracts/utils/StableMath.sol

sol2uml diff 0x07a450739cdd8Ac0a59C9d752ce915a86010658B .,node_modules --aFile contracts/harvest/Harvester.sol
Compared the "contracts/harvest/Harvester.sol" file with the "Harvester" contract with address 0x07a450739cdd8Ac0a59C9d752ce915a86010658B on mainnet
to local files under folders ".,node_modules"

match   @openzeppelin/contracts/token/ERC20/IERC20.sol
match   @openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol
match   @openzeppelin/contracts/utils/Address.sol
match   @openzeppelin/contracts/utils/math/Math.sol
match   @openzeppelin/contracts/utils/math/SafeMath.sol
match   contracts/governance/Governable.sol
match   contracts/harvest/BaseHarvester.sol
match   contracts/harvest/Harvester.sol
match   contracts/interfaces/balancer/IBalancerVault.sol
match   contracts/interfaces/IBasicToken.sol
match   contracts/interfaces/IOracle.sol
match   contracts/interfaces/IStrategy.sol
match   contracts/interfaces/IVault.sol
match   contracts/interfaces/uniswap/IUniswapV2Router02.sol
match   contracts/interfaces/uniswap/IUniswapV3Router.sol
match   contracts/strategies/ICurvePool.sol
match   contracts/token/OUSD.sol
match   contracts/utils/Helpers.sol
match   contracts/utils/Initializable.sol
match   contracts/utils/InitializableAbstractStrategy.sol
match   contracts/utils/InitializableERC20Detailed.sol
match   contracts/utils/StableMath.sol
match   contracts/vault/VaultStorage.sol

sol2uml diff 0x4De255cd196469752632e76CbAaAEeE0070D9A85 .,node_modules --aFile contracts/harvest/OETHHarvester.sol
Compared the "contracts/harvest/OETHHarvester.sol" file with the "OETHHarvester" contract with address 0x4De255cd196469752632e76CbAaAEeE0070D9A85 on mainnet
to local files under folders ".,node_modules"

match   @openzeppelin/contracts/token/ERC20/IERC20.sol
match   @openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol
match   @openzeppelin/contracts/utils/Address.sol
match   @openzeppelin/contracts/utils/math/Math.sol
match   @openzeppelin/contracts/utils/math/SafeMath.sol
match   contracts/governance/Governable.sol
match   contracts/harvest/BaseHarvester.sol
match   contracts/harvest/OETHHarvester.sol
match   contracts/interfaces/balancer/IBalancerVault.sol
match   contracts/interfaces/IBasicToken.sol
match   contracts/interfaces/IOracle.sol
match   contracts/interfaces/IStrategy.sol
match   contracts/interfaces/IVault.sol
match   contracts/interfaces/uniswap/IUniswapV2Router02.sol
match   contracts/interfaces/uniswap/IUniswapV3Router.sol
match   contracts/strategies/ICurvePool.sol
match   contracts/token/OUSD.sol
match   contracts/utils/Helpers.sol
match   contracts/utils/Initializable.sol
match   contracts/utils/InitializableAbstractStrategy.sol
match   contracts/utils/InitializableERC20Detailed.sol
match   contracts/utils/StableMath.sol
match   contracts/vault/VaultStorage.sol
```

**Comment by naddison36** [comment]
I've verified the following:
- immutable and storage variables in the deployed contracts
- Aura price was correct

The `strategistAddr` is not set on the `AuraWETHPriceFeed` contract but a follow up proposal will set this.

**Comment by sparrowDom** [comment]
Verifying proposal matches deploy script and that source code of deployed contracts match the deployed code:
- OETH harvester implementation is upgraded and impl contract code matches codebase
- OUSD harvester implementation is upgraded and impl contract code matches codebase
- dOETHRouter is set as new priceProvider and code in dOETHRouter matches our codebase
- reward token configs:
  - CRV (via triCRV) configured correctly
  - BAL configured correctly
  - AURA configured correctly
  - CVX configured correctly
  - CRV (OUSD) configured correctly
  - CVX (OUSD) configured correctly
  - COMP (OUSD) configured correctly
  - Aave (OUSD) configured correctly
- OETH dripper duration set to 14 days

Good to keep in mind that (OUSD)OracleRouter has codebase changes that are not going to be deployed.

**Review by naddison36 — APPROVED** [review]

**Comment by shahthepro** [comment]
> Good to keep in mind that (OUSD)OracleRouter has codebase changes that are not going to be deployed.

@sparrowDom There hasn't been any change to the OUSDOracleRouter (or OracleRouterBase) contract, other than splitting them up into separate files :)

File: contracts/utils/deploy.js [hunk 1]
```diff
@@ -346,14 +350,38 @@ const executeGovernanceProposalOnFork = async ({
...
+    let blocksToMine = deadline - currentBlock;
+
+    if (blocksToMine > 0) {
+      if (reduceQueueTime) {
+        blocksToMine = 10;
+        await setStorageAt(
+          governorFive.address,
+          slotKey,
+          // Make it queueable in 10 blocks
+          currentBlock + 10
```
Comment (sparrowDom) [inline contracts/utils/deploy.js:375]: nit: could also be: `currentBlock + blocksToMine`, since those 2 are related

**Review by naddison36 — APPROVED** [review]

**Pull request merged**